Checkstyle import control blog post. Fixes #149 (#206)
apache/accumulo-website: _posts/blog/2019-11-04-checkstyle-import-control.md

---
title: "Checking API use"
---

Accumulo follows [SemVer] across versions with the declaration of a public API.  Code not in the public API should be
considered unstable, at risk of changing between versions.  The public API packages are [listed on the website][api]
but may not be considered when an Accumulo user writes code.  This blog post explains how to make Maven
automatically detect usage of Accumulo code outside the public API.

The techniques described in this blog post only work for Accumulo 2.0 and later.  Do not use with 1.X versions.

## Checkstyle Plugin

First add the checkstyle Maven plugin to your pom.

```xml
<plugin>

>     <!-- This was added to ensure project only uses public API -->

Reviewer: ```suggestion
    <!-- This was added to ensure project only uses Accumulo's public API -->
```

## Import Control Configuration

Create the second file specified above, ```import-control.xml``` and copy the configuration below.  Make sure to replace
"<insert-your-package-name>" with the package name of your project.
```xml
<!DOCTYPE import-control PUBLIC
    "-//Checkstyle//DTD ImportControl Configuration 1.4//EN"
    "https://checkstyle.org/dtds/import_control_1_4.dtd">

<!-- This checkstyle rule is configured to ensure only use of Accumulo API -->
<import-control pkg="<insert-your-package-name>" strategyOnMismatch="allowed">
    <!-- allow this package -->
    <allow pkg="<insert-your-package-name>"/>
    <!-- API packages -->
    <allow pkg="org.apache.accumulo.core.client"/>
    <allow pkg="org.apache.accumulo.core.data"/>
    <allow pkg="org.apache.accumulo.core.security"/>
    <allow pkg="org.apache.accumulo.core.iterators"/>
    <allow pkg="org.apache.accumulo.minicluster"/>
    <allow pkg="org.apache.accumulo.hadoop.mapreduce"/>

    <!-- disallow everything else coming from accumulo -->
    <disallow pkg="org.apache.accumulo"/>
</import-control>
```
This file configures the ImportControl module to only allow packages that are declared public API.

> 

Reviewer: Could add a section like :

## Hold the line

Adding this to an existing project may expose usage of non public Accumulo API's.  It may take more time than is available to fix those at first, but do not let this discourage adding this plugin.  One possible way to proceed is to allow the currently used non-public APIs in a commented section of `import-control.xml` noting these are temporarily allowed until they can be removed.  This strategy prevents new usages of non-public APIs while allowing time to work on fixing the current usages of non public APIs.

---
title: "Checking API use"
---

Accumulo follows [SemVer] across versions with the declaration of a public API.  Code not in the public API should be

> considered unstable, at risk of changing between versions.  The packages included in the public API are [listed on the website][api]

Reviewer: ```suggestion
considered unstable, at risk of changing between versions.  The public API packages API are [listed on the website][api]
```

Author: I assume you meant: "The public API packages are [listed on the website][api]"

Reviewer: yes

> but may not always be considered when developing using Accumulo code.  This blog post explains how to setup a Maven project

Reviewer: ```suggestion
but may not be considered when an Accumulo user writes code.  This blog post explains how to make Maven
```

> to automatically detect when Accumulo 2.0 code used in the project is outside of the public API.

Reviewer: ```suggestion
automatically detect usage of Accumulo 2.0 code outside the public API.
```

> 

Reviewer: I think it would be really useful to add a sentence like the following.  This info is very important.

```
The techniques described in this blog post only work for Accumulo 2.0 and later.  Do not use these with 1.X versions because those version have `impl` packages inside API packages that must be excluded.  The needed exclusions for 1.X are not supported by import control.
```

Author: You are right, but I don't think we need to try to explain why.  I added a shortened version of this warning.

## Checkstyle Plugin

First add the checkstyle Maven plugin to your pom.

```xml
<plugin>
    <!-- This was added to ensure project only uses public API -->
    <groupId>org.apache.maven.plugins</groupId>
    <artifactId>maven-checkstyle-plugin</artifactId>
    <version>3.1.0</version>
    <configuration>
      <configLocation>checkstyle.xml</configLocation>

>     </configuration>

Reviewer 2: There are different implications for putting this config here rather than in the execution section below. This config will be used even for manual runs of the plugin. I'm not sure that's the behavior we want to encourage people to use. Maybe move this config to the execution section below.

Reviewer: I am learning more about Maven from these comments.

Author: > This config will be used even for manual runs of the plugin.

What is wrong with that?  I would assume if this gets added to the checkstyle plugin, it would always run when checkstyle runs.

Reviewer 2: > What is wrong with that?

It's not necessarily "wrong"... but it can make it more difficult to run the checkstyle plugin with different configuration in separate executions. Moving the `<configuration/>` element and its contents to inside the `<execution/>` element will make it so that execution is specifically configured to do the import control checks, but it wouldn't affect any other executions of the checkstyle plugin the user may already have configured (or might want to configure later), such as additional checks in a separate execution defined in a profile, or executions run manually on the command-line, or independent executions in sibling modules in a multi-module project. In the case of sibling modules, it might be useful for the user to have separate import-control configs, each with their own `<import-control pkg="my.package.name.example">` settings for that specific module, rather than have one massive config file for all modules.

    <dependencies>
      <dependency>
        <groupId>com.puppycrawl.tools</groupId>
        <artifactId>checkstyle</artifactId>
        <version>8.23</version>
      </dependency>

>     </dependencies>

Reviewer 2: This section is optional and could be removed in the example, unless we need to override the version of checkstyle to get a specific version for this functionality to work. Otherwise, most people are probably just going to use the version used by the plugin. Removing this makes it more likely that the user will have a good experience here, because some versions of the checkstyle dependency may not work as well as those tested by the plugin developers.

Author: Good point.  I will remove it.

```
The plugin version is the latest at the time of this post.  For more information see the website for
the [Apache Maven Checkstyle Plugin][plugin].  The configuration above adds the plugin to ```check``` execution goal
so it will always run with your build.  

Create the configuration file specified above: ```checkstyle.xml```

### checkstyle.xml

```xml
<!DOCTYPE module PUBLIC "-//Puppy Crawl//DTD Check Configuration 1.3//EN" "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
  <property name="charset" value="UTF-8"/>
  <module name="TreeWalker">
    <!--check that only Accumulo public APIs are imported-->
    <module name="ImportControl">
      <property name="file" value="import-control.xml"/>
    </module>
  </module>
</module>
```
This file sets up the ImportControl module.

## Import Control Configuration

Create the second file specified above, ```import-control.xml``` and copy the configuration below:
```xml
<!DOCTYPE import-control PUBLIC
    "-//Checkstyle//DTD ImportControl Configuration 1.4//EN"
    "https://checkstyle.org/dtds/import_control_1_4.dtd">

<!-- This checkstyle rule is configured to ensure only use of Accumulo API -->

> <import-control pkg="org.apache.accumulo.testing" strategyOnMismatch="allowed">

Reviewer 2: Is this `pkg` value correct for users? It looks like this will only enforce import control over classes in our own testing code. It would be useful to explain to the reader what they must do to apply it to their own code.

And, is it worth mentioning what the `strategyOnMismatch` does? I'm kind of curious myself.

Author: You are correct, that should be whatever the package of their project is.

Author: > And, is it worth mentioning what the `strategyOnMismatch` does? I'm kind of curious myself.

I don't remember exactly why this was needed so I didn't mention it.

Author: I think this strategy is needed because the module is so explicit.  Without allowing a mismatch, any import not allowed will get flagged, like even java imports.

    <!-- allow this package -->
    <allow pkg="org.apache.accumulo.testing"/>
    <!-- API packages -->
    <allow pkg="org.apache.accumulo.core.client"/>
    <allow pkg="org.apache.accumulo.core.data"/>
    <allow pkg="org.apache.accumulo.core.security"/>
    <allow pkg="org.apache.accumulo.core.iterators"/>
    <allow pkg="org.apache.accumulo.minicluster"/>
    <allow pkg="org.apache.accumulo.hadoop.mapreduce"/>

    <!-- SPI package -->

>     <allow pkg="org.apache.accumulo.core.spi"/>

Reviewer: SPI is not currently public API.. not sure if we should include it in this post.

Author: Yeah I wasn't sure either.  We don't list it on the site so I'll remove it.
